proposals/20200828-structured-exception-handling.md

```diff
@@ -0,0 +1,214 @@
+    - [npm, /usr/node/bin]
+  - container_writer:
+    - [docker.io/alpine, /usr/libexec/alpine]
+```
+
+The name of the exception links it to the rule.
+
+A rule exception applies if for a given event, the fields in a rule.exception match all of the values in some exception.item. For example, if a program `apk` writes to a file below `/usr/lib/alpine`, the rule will not trigger, even if the condition is met.
+
+Append will be supported for exception objects. If append: is true, the items in the second definition will be added to the items in the earlier definition. For example, adding:
+
+```
+- exception: Write below binary dir
+  items:
+  - container_writer:
+    - [docker.io/golang-alpine, /usr/libexec/go]
+  append: true
+```
+
+Would add a second container_writer exception.
+
+### Implementation
+
+Each exception can be thought of as an implicit "and not (field1=val1 and field2=val2 and...)" appended to the rule's condition. In practice, that's how exceptions will be implemented.
+
+When a rule is parsed, the original condition will be wrapped in an extra layer of parentheses and all exception values will be appended to the condition. For example, using the example above, the resulting condition will be:
+
+```
+(<Write below binary dir condition>) and not ((proc.name=apk and fd.directory=/usr/lib/alpine) or (proc.name=npm and fd.directory=/usr/node/bin) or (container.image.repository=docker.io/alpine and fd.directory=/usr/libexec/alpine))
+```
+
+The exceptions are effectively syntatic sugar that allows expressing sets of exceptions in a concise way.
+
+### Advantages
+
+Adding Exception objects as described here has several advantages:
+
+* All rules will implicitly support exceptions. A rule writer doesn't need to define a user_known_xxx macro and add it to the condition.
+* The rule writer has some controls on what defines a valid exception. The rule author knows best what is a good exception, and can define the fields that make up the exception.
+* With this approach, it's much easier to add and manage multiple sets of exceptions from multiple sources. You're just combining lists of tuples of filtercheck field values.
+
+## Backwards compatibility
+
+This approach does not remove the ability to append to exceptions nor the existing use of user_xxx macros to define exceptions to rules. It only provides an additional way to express exceptions. Hopefully, we can migrate existing exceptions to use this approach, but there isn't any plan to make wholesale rules changes as a part of this.
+
+This approach is for the most part backwards compatible with older falco releases. To implement exceptions, we'll add a preprocessing element to rule parsing. The main falco engine is unchanged.
+
+However, there are a few changes we'll have to make to falco rules file parsing:
+
+* Currently, falco will reject files with top level `-exception` objects. We'll probably want to make a one-time change to Falco to allow unknown top level objects.
```

Comment:
```suggestion
* Currently, Falco will reject files with top-level `-exception` objects. We'll probably want to make a one-time change to Falco to allow unknown top-level objects.
```

Reply:
Should it be `- exception` instead of `-exception`?

Reply:
I was just trying to call out the exception object as a top level list item. I can reword it.